src/hera/workflows/resource_base.py

```diff
@@ -0,0 +1,186 @@
+"""Define the base classes used to produce k8s CRDs."""
+from __future__ import annotations
+
+from collections import ChainMap
+from typing import Any, Callable, ClassVar, Dict, List, Optional, Type, Union
+
+try:
+    from inspect import get_annotations  # type: ignore
+except ImportError:
+    from hera.workflows._inspect import get_annotations  # type: ignore
+from pathlib import Path
+
+try:
+    from typing import Annotated, get_args, get_origin  # type: ignore
+except ImportError:
+    from typing_extensions import Annotated, get_args, get_origin  # type: ignore
+
+try:
+    from typing import Self  # type: ignore
+except ImportError:
+    from typing_extensions import Self  # type: ignore
+
+from hera import _yaml
+from hera.shared import BaseMixin
+from hera.shared._pydantic import PydanticBaseModel
+
+
+class ModelMapper:
+    """Maps a model attribute to the attribute path of the produced resource.
+
+    For example `foo: MopdelMapper("metadata.name")` would map the annotated object's
+    `foo` attribute to the "metadata" -> "name" field on the final k8s resource.
+    """
+
+    def __init__(self, model_path: str, hera_builder: Optional[Callable] = None):
+        """Accept the `model_path` to map to."""
+        self.model_path = None
+        self.builder = hera_builder
+
+        if not model_path:
+            # Allows overriding parent attribute annotations to remove the mapping
+            return
+
+        self.model_path = model_path.split(".")
+
+    @classmethod
+    def build_model(
+        cls,
+        hera_class: Type[ResourceBase],
+        hera_obj: ResourceBase,
+        model: PydanticBaseModel,
+        *,
+        traverse_mro: bool = True,
+    ):
+        """Builds an instance of a `ResourceBase` into the destination model type."""
+        assert isinstance(hera_obj, ResourceBase), type(hera_obj)
+
+        for attr, mapper in hera_class._iter_model_mappers(traverse_mro=traverse_mro):
+            # Value comes from builder function if it exists on hera_obj, otherwise directly from the attr
+            value = (
+                getattr(hera_obj, mapper.builder.__name__)() if mapper.builder is not None else getattr(hera_obj, attr)
+            )
+            if value is not None:
+                _set_model_attr(model, mapper.model_path, value)
+
+        return model
+
+
+class ResourceBase(BaseMixin):
+    """Base class for top-level Hera CRDs.
+
+    Subclasses should supply a `mapped_model` attribute, which is the Pydantic model
+    which represents the final k8s resource.
+
+    Note, this class will validate that the annotated attributes map to real fields on the `mapped_model`.
+    """
+
+    mapped_model: ClassVar[Type[PydanticBaseModel]]
```

Comment:
I perhaps think at least api_version,** and kind should be defined at the base-class level. It's not a huge deal, but it would save 2 lines in each of the existing subclasses, and generally simplify any external implementers to only those fields which are actually unique to that CRD

Although it will probably take some extra work specifically because of how i handled CronWorfklow being weird.